File: pyftsm/test/ftsm_failed_configuration_test.py
import time
from pyftsm.ftsm_base import FTSMBase, FTSMTransitions

class Component(FTSMBase):
    def __init__(self, name, dependencies, max_recovery_attempts=1):
        super(Component, self).__init__(name, dependencies, max_recovery_attempts)

    def init(self):
        print('Initialising')
        return FTSMTransitions.INITIALISED

    def configuring(self):
        print('Configuring')
        time.sleep(0.2)
        return FTSMTransitions.RETRY_CONFIG

    def ready(self):
        print('Ready')
        time.sleep(0.25)
        return FTSMTransitions.RUN

    def running(self):
        print('Running')
        time.sleep(1.)
        return FTSMTransitions.DONE

    def recovering(self):
        print('Recovering')
        time.sleep(0.5)
        return FTSMTransitions.DONE_RECOVERING

if __name__ == '__main__':
    component = Component('component_that_cannot_configure', ['component_1', 'component_2'], 3)
    try:
        component.run()
        while component.is_running:
            time.sleep(0.1)
    except (KeyboardInterrupt, SystemExit):
        print('{0} interrupted; exiting...'.format(component.name))
        component.stop()

File: cppftsm/src/test/ftsm_failed_configuration_test.cpp
#include "ftsm.h"

#include <csignal>
#include <string>
#include <vector>
#include <thread>
#include <chrono>

namespace ftsm_tests
{
    class FailedConfigComponent : public ftsm::FTSM
    {
    public:
        FailedConfigComponent(std::string name, std::vector<std::string> dependencies, int max_recovery_attempts=1)
         : ftsm::FTSM(name, dependencies, max_recovery_attempts) { }

        virtual std::string init()
        {
            std::cout << "initialising..." << std::endl;
            return ftsm::FTSMTransitions::INITIALISED;
        }

        virtual std::string configuring()
        {
            std::cout << "configuring..." << std::endl;
            std::this_thread::sleep_for(std::chrono::milliseconds(200));
            return ftsm::FTSMTransitions::RETRY_CONFIG;
        }

        virtual std::string ready()
        {
            std::cout << "waiting..." << std::endl;
            std::this_thread::sleep_for(std::chrono::milliseconds(250));
            return ftsm::FTSMTransitions::RUN;
        }

        virtual std::string running()
        {
            std::cout << "running..." << std::endl;
            std::this_thread::sleep_for(std::chrono::milliseconds(1000));
            return ftsm::FTSMTransitions::DONE;
        }

        virtual std::string recovering()
        {
            std::cout << "recovering..." << std::endl;
            std::this_thread::sleep_for(std::chrono::milliseconds(500));
            return ftsm::FTSMTransitions::DONE_RECOVERING;
        }
    };
}

ftsm_tests::FailedConfigComponent component("component_that_cannot_configure", {"component1", "component2"}, 3);

void checkTermination(int signal)
{
    component.stop();
}

int main()
{
    component.run();
    signal(SIGINT, checkTermination);
    while (component.is_running)
    {
        std::this_thread::sleep_for(std::chrono::milliseconds(100));
    }

    if (component.is_alive)
    {
        component.stop();
    }
    return 0;
}
